core/types/transactions/message.go

```diff
@@ -27,18 +41,58 @@ type CallMessage struct {
 
 	// Sender is the public key of the sender
 	Sender types.HexBytes `json:"sender"`
+
+	// Signature is the sender's signature of the serialized call body. This is
+	// only set when using authenticated call RPCs, in which case the the
+	// Challenge field of the call body is also set.
+	Signature *auth.Signature `json:"signature"`
+}
+
+const callMsgToSignTmplV0 = `Kwil view call.
+
+DBID: %s
+Method: %s
+Digest: %x
+Challenge: %x
+`
+
+func CallSigText(dbid, action string, payload []byte, challenge []byte) string {
+	digest := sha256.Sum256(payload)
+	return fmt.Sprintf(callMsgToSignTmplV0, dbid, action, digest[:20], challenge)
 }
 
-// CreateCallMessage creates a new call message from a ActionCall payload.
-func CreateCallMessage(payload *ActionCall) (*CallMessage, error) {
-	bts, err := payload.MarshalBinary()
+// CreateCallMessage creates a new call message from a ActionCall payload. If a
+// signer is provided, the sender and authenticator type are set. If a challenge
+// is also provided, it will also sign a serialization of the request that
+// includes the challenge for replay protection. Thus, if a challenge is
+// provided, a signer must also be provided.
+func CreateCallMessage(ac *ActionCall, challenge []byte, signer auth.Signer) (*CallMessage, error) {
+	bts, err := ac.MarshalBinary()
 	if err != nil {
 		return nil, err
 	}
 
-	return &CallMessage{
+	msg := &CallMessage{
 		Body: &CallMessageBody{
-			Payload: bts,
+			Payload:   bts,
+			Challenge: challenge,
 		},
-	}, nil
+	}
+
+	if signer != nil { // for @caller
+		msg.AuthType = signer.AuthType()
+		msg.Sender = signer.Identity()
+	}
+
+	if len(challenge) > 0 {
+		sigText := CallSigText(ac.DBID, ac.Action, bts, challenge)
+		sig, err := signer.Sign([]byte(sigText))
```

Comment:
`signer` could be nil here

Reply:
I don't think so. I assume you're referring to `signer` sometimes being nil if it was not configured in a `client`. However, I feel like any checks for that should be done at a higher level, since this is simply a function in the transactions package.

Reply:
Yeah, this would be vanilla misuse of a the function.  I've opted for docs:

```go
// CreateCallMessage creates a new call message from a ActionCall payload. If a
// signer is provided, the sender and authenticator type are set. If a challenge
// is also provided, it will also sign a serialization of the request that
// includes the challenge for replay protection. Thus, if a challenge is
// provided, a signer must also be provided.
```